sky/server/server.py

```diff
@@ -1756,57 +1758,119 @@ async def kubernetes_pod_ssh_proxy(websocket: fastapi.WebSocket,
-                while True:
-                    data = await reader.read(1024)
-                    if not data:
-                        if not websocket_closed:
-                            logger.warning('SSH connection to pod is '
-                                           'disconnected before websocket '
-                                           'connection is closed')
-                            nonlocal ssh_failed
+                ssh_socket.settimeout(0.1)  # Short timeout for non-blocking check
+                while not shutdown_event.is_set():
+                    try:
+                        # Read from SSH socket (blocking, no yielding)
+                        data = ssh_socket.recv(1024)
+                        if not data:
+                            if not websocket_closed:
+                                logger.warning('SSH connection to pod disconnected')
+                                ssh_failed = True
+                            break
+                            
+                        # Send to WebSocket (this needs to be done in the main loop)
+                        future = asyncio.run_coroutine_threadsafe(
+                            websocket.send_bytes(data), loop)
+                        future.result(timeout=1.0)  # Longer timeout for send
+                        
+                    except socket.timeout:
+                        continue  # Check shutdown_event and try again
+                    except Exception as e:
+                        if not shutdown_event.is_set():
+                            logger.error(f'SSH to WebSocket error: {e}')
                             ssh_failed = True
                         break
-                    await websocket.send_bytes(data)
-            except Exception:  # pylint: disable=broad-except
-                pass
-            try:
-                await websocket.close()
-            except Exception:  # pylint: disable=broad-except
-                # The websocket might has been closed by the client.
-                pass
-
-        await asyncio.gather(websocket_to_ssh(), ssh_to_websocket())
+            except Exception as e:
+                logger.error(f'SSH to WebSocket thread error: {e}')
+                ssh_failed = True
+            finally:
+                shutdown_event.set()
+                # Close WebSocket from main thread
+                try:
+                    future = asyncio.run_coroutine_threadsafe(
+                        websocket.close(), loop)
+                    future.result(timeout=1.0)
+                except Exception:
+                    pass  # WebSocket might already be closed
+                
+        # Start both threads
+        ws_to_ssh_thread = threading.Thread(target=websocket_to_ssh_thread, 
+                                           name="ws-to-ssh")
+        ssh_to_ws_thread = threading.Thread(target=ssh_to_websocket_thread,
+                                           name="ssh-to-ws")
```

Comment:
why not context_utils.to_thread, but a separate thread? Would there be a trade off here, e.g., it becomes unbounded, and may lead to OOM/oversubscriptoin.